cmd/crank/beta/validate/validate_test.go

```diff
@@ -1338,3 +1340,99 @@ func TestValidateResources(t *testing.T) {
 		})
 	}
 }
+
+func TestValidateUnknownFields(t *testing.T) {
+	type args struct {
+		mr  map[string]interface{}
+		sch *schema.Structural
+	}
+	type want struct {
+		errField string
+		errMsg   string
+		hasError bool
+	}
+	cases := map[string]struct {
+		reason string
+		args   args
+		want   want
+	}{
+		"UnknownFieldPresent": {
+			reason: "Should detect unknown fields in the resource and return an error",
+			args: args{
+				mr: map[string]interface{}{
+					"apiVersion": "test.org/v1alpha1",
+					"kind":       "Test",
+					"metadata": map[string]interface{}{
+						"name": "test-instance",
+					},
+					"spec": map[string]interface{}{
+						"replicas":     3,
+						"unknownField": "should fail", // This field is not defined in the CRD schema
+					},
+				},
+				sch: &schema.Structural{
+					Properties: map[string]schema.Structural{
+						"spec": {
+							Properties: map[string]schema.Structural{
+								"replicas": {
+									Generic: schema.Generic{Type: "integer"},
+								},
+							},
+						},
+					},
+				},
+			},
+			want: want{
+				errField: "spec.unknownField",
+				errMsg:   "unknown field: \"unknownField\"",
+				hasError: true,
+			},
+		},
+		"UnknownFieldNotPresent": {
+			reason: "Should not return an error when no unknown fields are present",
+			args: args{
+				mr: map[string]interface{}{
+					"apiVersion": "test.org/v1alpha1",
+					"kind":       "Test",
+					"metadata": map[string]interface{}{
+						"name": "test-instance",
+					},
+					"spec": map[string]interface{}{
+						"replicas": 3, // No unknown fields
+					},
+				},
+				sch: &schema.Structural{
+					Properties: map[string]schema.Structural{
+						"spec": {
+							Properties: map[string]schema.Structural{
+								"replicas": {
+									Generic: schema.Generic{Type: "integer"},
+								},
+							},
+						},
+					},
+				},
+			},
+			want: want{
+				hasError: false,
+			},
+		},
+	}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			errs := validateUnknownFields(tc.args.mr, tc.args.sch)
+			if tc.want.hasError {
```

Comment:
great to see a unit test get added here too @enesonus, so we can have confidence that your new logic works now and into the future as code gets updated ✅ 

I thought of an slightly different approach for validating the results of this function, by just examining the entire `field.ErrorList`.  That way, you can remove a lot of the extra testing checks here and just let `cmp.Diff()` do the heavy lifting.

What do you think of this approach? Feel free to copy it into your PR here if you think its an improvement 😉 

https://github.com/jbw976/crossplane/commit/0cf9a97167a067504d36e76cde29c84f4368bd47

Reply:
awesome, thank you for evaluating and incorporating the suggestion @enesonus!